### PR TITLE
Microservice authoring guidelines

### PR DESCRIPTION
**Pull request by tiagojsag**
https://gfw.atlassian.net/browse/GTC-222

### CODE REVIEW

File: source/includes/dev/_microservice.md [hunk 1]
```diff
@@ -1,17 +1,109 @@
-# Microservice
+# Microservice development guide
 
-This section of the docs covers the details of a microservice.
+In this chapter, we'll cover additional details that you, as a RW API developer, should keep in mind when developing your microservice. We'll focus not only on the technical requirements you need to meet for your microservice to communicate with the remaining RW API internal components, bur also discuss the policy surrounding development for the RW API, as a way to achieve a certain degree of consistency across a naturally heterogeneous microservice-based system.
```
Comment (ikas) [inline source/includes/dev/_microservice.md]:
```suggestion
In this chapter, we'll cover additional details that you, as a RW API developer, should keep in mind when developing your microservice. We'll focus not only on the technical requirements you need to meet for your microservice to communicate with the remaining RW API internal components, but also discuss the policy surrounding development for the RW API, as a way to achieve a certain degree of consistency across a naturally heterogeneous microservice-based system. 
```

File: source/includes/dev/_microservice.md [hunk 2]
```diff
@@ -1,17 +1,109 @@
...
+This section is completely implementation specific and opinionated, and does not reflect any technical requirement of the API. However, as all nodejs microservices use this architecture, we will cover it here as it's useful for new developers. Other microservice implementations in other languages will have different architectures, and you can also implement your own microservice using nodejs using a totally different architecture. Take this whole section as information/suggestion rather than as a ruleset.
 </aside>
 
+Nodejs microservices are based on the [Koa](https://koajs.com/) framework for nodejs. To understand the following code snippets, we assume you are familiar with the basics of the framework, like how routes are declared and handled, and what middleware are and how they work. You should also be somewhat familiar with tools like [npm](https://www.npmjs.com/), [mongo](https://www.mongodb.com/) and [mongoose](https://mongoosejs.com/), [Jenkins CI](https://jenkins.io/), [docker](https://www.docker.com/) and [docker-compose](https://docs.docker.com/compose/)
```
Comment (ikas) [inline source/includes/dev/_microservice.md]:
```suggestion
Nodejs microservices are based on the [Koa](https://koajs.com/) framework for nodejs. To understand the following code snippets, we assume you are familiar with the basics of the framework, like how routes are declared and handled, what middleware is, and how it works. You should also be somewhat familiar with tools like [npm](https://www.npmjs.com/), [mongo](https://www.mongodb.com/) and [mongoose](https://mongoosejs.com/), [Jenkins CI](https://jenkins.io/), [docker](https://www.docker.com/) and [docker-compose](https://docs.docker.com/compose/)
```

File: source/includes/dev/_microservice.md [hunk 3]
```diff
@@ -1,17 +1,109 @@
...
+router.post('/:dataset/recover', authorizationRecover, DatasetRouter.recover);
+
+router.get('/:dataset', DatasetRouter.get);
+router.get('/:dataset/verification', DatasetRouter.verification);
+router.patch('/:dataset', validationMiddleware, authorizationMiddleware, DatasetRouter.update);
+router.delete('/:dataset', authorizationMiddleware, DatasetRouter.delete);
+router.post('/:dataset/clone', validationMiddleware, authorizationMiddleware, DatasetRouter.clone);
+```
+
+In here you'll find the already existing routes. As you can see from the rather explicit syntax, you need to call the method that matches the desired HTTP verb on the `router` object, and pass it a variable number of arguments - more on this in the next section. One thing to keep in mind is that all the routes in a file are typically prefixed, as defined in the `router` object declaration.
```
Comment (ikas) [inline source/includes/dev/_microservice.md:104]: I'd add here a statement saying that we use `koa-router` for the routers and a link to those docs, for more info: https://github.com/ZijianHe/koa-router

File: source/includes/dev/_microservice.md [hunk 4]
```diff
@@ -126,14 +218,81 @@ As explained above, although the call is ultimately to another microservice, the
...
+
+Its worth noting that these container are set up in a way that allows using them to both run the microservice itself, or their tests. This will be useful further ahead when we review the testing approach you should use when writing microservices.
+
+## Data layer
+
+Many microservices require the ability to store data to perform their function. The RW API has several data storage tools available to you, in case you need to store information to run your service.
+
+**Warning**: microservices run on ephemeral containers managed by Kubernetes, and often in multiple parallel instances, so do not rely on storing data on the filesystem, unless you know there's something like a [Kubernetes' persistent volume](https://kubernetes.io/docs/concepts/storage/persistent-volumes/) to back it up.
+
+When accessing these tools, there are a few things you should keep in mind:
```
Comment (ikas) [inline source/includes/dev/_microservice.md:235]: List not correctly formatted, missing a line break before the items

File: source/includes/dev/_microservice.md [hunk 5]
```diff
@@ -126,14 +218,81 @@ As explained above, although the call is ultimately to another microservice, the
...
+
+The RW API has a system-wide HTTP cache that you may use to cache your requests, improving scalability and response times. This cache is based on [Fastly](https://www.fastly.com/), and you can browse its documentation if you are looking for a specific detail on its behavior. For most common use cases, you just need to keep in mind the following:
+
+- Only responses with codes 200, 203, 300 and 410 are cached, and the default cache TTL is 3 days.
+- GET responses for `/auth` endpoints are never cached.
+- GET responses for `/query` or `/fields` endpoints are cached for 2 days.
+- Your microservice can use the `cache` response header to tag a cache entry. This, in itself, has no functional impact. Example [here](https://github.com/resource-watch/dataset/blob/47ad8b9509b97803d7f484549908e72ecaa98467/app/src/routes/api/v1/dataset.router.js#L396).
+- Your microservice can use the `uncache` header to purge cache entries matching a given tag, as set in the previous point. Example [here](https://github.com/resource-watch/dataset/blob/47ad8b9509b97803d7f484549908e72ecaa98467/app/src/routes/api/v1/dataset.router.js#L462).
+
+## Logging
```
Comment (ikas) [inline source/includes/dev/_microservice.md:290]: I'd add, somewhere in this section, that you can (**and should!**) configure the error reporting level in most MSs using the `LOGGER_LEVEL` env variable.

File: source/includes/dev/_microservice.md [hunk 6]
```diff
@@ -142,97 +301,225 @@ const logger = require('logger');
...
+Different microservices and endpoint will have different requirements when it comes to testing, but the great majority of endpoints can be tested using simple variations of these steps. There are some additional considerations you should take into account when testing:
 
-To add a new endpoint, here's the short tasklist you have to tackle:
+- The example above creates an actual dataset, meaning a MongoDB (or equivalent mocks) need to exist. For MongoDB specifically, our approach so far has been to use a real MongoDB instance, and running the tests on a separate collection (´dataset-tests´ for example), aiming for isolation. Other microservices (for example, those relying on Elasticsearch) use mocks instead. Mocking usually leads to faster execution times, but can be troublesome to properly code. Use whichever alternative is best for you, and refer to the [Data layer](#data-layer) section for examples of microservices that use (and test with) different tools.
```
Comment (ikas) [inline source/includes/dev/_microservice.md]:
```suggestion
- The example above creates an actual dataset, meaning a MongoDB (or equivalent mocks) need to exist. For MongoDB specifically, our approach so far has been to use a real MongoDB instance, and running the tests on a separate database (´dataset-tests´ for example), aiming for isolation. Other microservices (for example, those relying on Elasticsearch) use mocks instead. Mocking usually leads to faster execution times, but can be troublesome to properly code. Use whichever alternative is best for you, and refer to the [Data layer](#data-layer) section for examples of microservices that use (and test with) different tools.
```

File: source/includes/dev/_microservice.md [hunk 7]
```diff
@@ -142,97 +301,225 @@ const logger = require('logger');
...
+- Nock has a [feature that blocks all HTTP requests](https://github.com/nock/nock#enabledisable-real-http-requests), which is useful to ensure your code or tests are not relying on an external service without you being aware - just be sure to whitelist your own IP, otherwise the HTTP call your test makes to your microservice will fail too.
+- Individual tests should run in isolation, and without assuming order. For example, running a test that first tests an insert, and then using the inserted element to test a delete would be a bad practice. Instead, your insert test should clean up its data once it's done, and the delete test should prepopulate the database before actually trying to delete it. A corollary of this is that you should be able to run your tests multiple times, back-to-back, without that affecting the results.
```
Comment (ikas) [inline source/includes/dev/_microservice.md]: I'd summarize this last bullet as "Always make sure your tests are idempotent." :grin:

**Review by ikas — APPROVED** [review]
Some minor changes, but overall looks really good, great job :clap:

**Review by ethanroday — APPROVED** [review]

**Pull request merged**